server/gc/fsck.c

```diff
@@ -175,160 +309,152 @@ recover_corrupted_repo_head (char *repo_id)
 
     for (temp_list = commit_list; temp_list; temp_list = temp_list->next) {
         temp_commit = temp_list->data;
+        io_error = FALSE;
+
+        if (!fsck_verify_seafobj (repo->store_id, 1, temp_commit->root_id,
+                                  &io_error, VERIFY_DIR, clean)) {
+            if (io_error) {
+                seaf_warning ("IO error, stop to run fsck for repo %.8s.\n",
+                              repo_id);
+                seaf_repo_unref (repo);
+                repo = NULL;
+                goto out;
+            }
+            // fs object of this commit is curropted,
+            // continue to verify next
+            continue;
+        }
 
         branch = seaf_branch_new ("master", repo_id, temp_commit->commit_id);
         if (branch == NULL) {
-            continue;
-        }
-        repo = seaf_repo_new (repo_id, NULL, NULL);
-        if (repo == NULL) {
-            seaf_branch_unref (branch);
-            continue;
+            seaf_warning ("Out of memory, stop to run fsck for repo %.8s.\n",
+                          repo_id);
+            seaf_repo_unref (repo);
+            repo = NULL;
+            goto out;
         }
         repo->head = branch;
         seaf_repo_from_commit (repo, temp_commit);
-        vinfo = seaf_repo_manager_get_virtual_repo_info (seaf->repo_mgr, repo_id);
-        if (vinfo) {
-            repo->is_virtual = TRUE;
-            memcpy (repo->store_id, vinfo->origin_repo_id, 36);
-        } else {
-            repo->is_virtual = FALSE;
-            memcpy (repo->store_id, repo->id, 36);
-        }
-        seaf_virtual_repo_info_free (vinfo);
 
-        res.repo = repo;
-        rc = seaf_fs_manager_traverse_tree (seaf->fs_mgr,
-                                            repo->store_id,
-                                            repo->version,
-                                            temp_commit->root_id,
-                                            fs_callback,
-                                            &res, FALSE);
+        char time_buf[64];
+        strftime(time_buf, 64, "%Y-%m-%d %H:%M:%S", localtime((time_t *)&temp_commit->ctime));
+        seaf_message ("Find available commit %.8s(created at %s) for repo %.8s.\n",
+                      temp_commit->commit_id, time_buf, repo_id);
 
-        if (rc < 0) {
-            seaf_repo_unref (repo);
-        } else {
-            break;
-        }
-    }
-
-    if (rc < 0) {
-        seaf_warning ("Failed to fix head commit of repo %.8s.\n", repo_id);
-    } else {
-        // create new head commit, and set it's parent commit as latest avaliable commit
-        temp_commit = cre_commit_from_parent (repo_id, temp_commit);
-        if (temp_commit) {
-            seaf_branch_set_commit (repo->head, temp_commit->commit_id);
-            // in case of branch col miss, using add_branch instead of update_branch
+        if (clean) {
             if (seaf_branch_manager_add_branch (seaf->branch_mgr, repo->head) < 0) {
```

Comment:
Don't change database at this point. Move this code to check_and_recover_repo().